### ISSUE TITLE
test_ignore_dead_nodes_for_replace_option: Startup failed: std::runtime_error (Failed to parse node list: {127.0.55.6, 127.0.55.7}: invalid node=127.0.55.6: std::runtime_error (Host inet address 127.0.55.6 not found in the cluster))

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#14487]
https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/294/artifact/logs-full.release.019/1688369920023_repair_based_node_operations_test.py%3A%3ATestRepairBasedNodeOperations%3A%3Atest_ignore_dead_nodes_for_replace_option/node8.log
```
Scylla version 5.4.0~dev-0.20230703.1ab2bb69b8a6 with build-id 1054adfa55f238441c3044e6213fd02d31d43279 starting ...
command used: "/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-nvlixxhx/test/node8/bin/scylla --options-file /jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-nvlixxhx/test/node8/conf/scylla.yaml --log-to-stdout 1 --abort-on-seastar-bad-alloc --abort-on-lsa-bad-alloc 1 --abort-on-internal-error 1 --enable-repair-based-node-ops true --ignore-dead-nodes-for-replace 127.0.55.6,127.0.55.7 --api-address 127.0.55.8 --collectd-hostname d332c648e72c.node8 --smp 2 --memory 1024M --developer-mode true --default-log-level info --collectd 0 --overprovisioned --prometheus-address 127.0.55.8 --replace-node-first-boot d97b63b2-72f5-4640-bba8-893420adf609 --unsafe-bypass-fsync 1 --kernel-page-cache 1 --commitlog-use-o-dsync 0 --max-networking-io-control-blocks 1000"
parsed command line options: [options-file: /jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-nvlixxhx/test/node8/conf/scylla.yaml, log-to-stdout, (positional) 1, abort-on-seastar-bad-alloc, abort-on-lsa-bad-alloc: 1, abort-on-internal-error: 1, enable-repair-based-node-ops: true, ignore-dead-nodes-for-replace: 127.0.55.6,127.0.55.7, api-address: 127.0.55.8, collectd-hostname, (positional) d332c648e72c.node8, smp, (positional) 2, memory, (positional) 1024M, developer-mode: true, default-log-level, (positional) info, collectd, (positional) 0, overprovisioned, prometheus-address: 127.0.55.8, replace-node-first-boot: d97b63b2-72f5-4640-bba8-893420adf609, unsafe-bypass-fsync, (positional) 1, kernel-page-cache, (positional) 1, commitlog-use-o-dsync: 0, max-networking-io-control-blocks, (positional) 1000]

INFO  2023-07-03 07:38:26,583 [shard 0] init - seeds={127.0.55.1, 127.0.55.2, 127.0.55.3, 127.0.55.4, 127.0.55.5, 127.0.55.6, 127.0.55.7}, listen_address=127.0.55.8, broadcast_address=127.0.55.8

INFO  2023-07-03 07:38:26,662 [shard 0] storage_service - entering STARTING mode
INFO  2023-07-03 07:38:26,662 [shard 0] storage_service - Loading persisted ring state
INFO  2023-07-03 07:38:26,664 [shard 0] storage_service - initial_contact_nodes={127.0.55.7, 127.0.55.6, 127.0.55.5, 127.0.55.4, 127.0.55.3, 127.0.55.2, 127.0.55.1}, loaded_endpoints={}, loaded_peer_features=0
INFO  2023-07-03 07:38:26,664 [shard 0] storage_service - Gathering node replacement information for d97b63b2-72f5-4640-bba8-893420adf609/0000:0000:0000:0000:0000:0000:0000:0000
INFO  2023-07-03 07:38:26,664 [shard 0] storage_service - Checking remote features with gossip
INFO  2023-07-03 07:38:26,664 [shard 0] gossip - Gossip shadow round started with nodes={127.0.55.7, 127.0.55.6, 127.0.55.5, 127.0.55.4, 127.0.55.3, 127.0.55.2, 127.0.55.1}
WARN  2023-07-03 07:38:26,665 [shard 0] gossip - Node 127.0.55.7 is down for get_endpoint_states verb
WARN  2023-07-03 07:38:26,665 [shard 0] gossip - Node 127.0.55.6 is down for get_endpoint_states verb
WARN  2023-07-03 07:38:26,665 [shard 0] gossip - Node 127.0.55.5 is down for get_endpoint_states verb
INFO  2023-07-03 07:38:26,666 [shard 0] gossip - Gossip shadow round finished with nodes_talked={127.0.55.2, 127.0.55.1, 127.0.55.3, 127.0.55.4}
INFO  2023-07-03 07:38:26,666 [shard 0] gossip - Feature check passed. Local node 127.0.55.8 features = {AGGREGATE_STORAGE_OPTIONS, ALTERNATOR_TTL, CDC, CDC_GENERATIONS_V2, COLLECTION_INDEXING, COMPUTED_COLUMNS, CORRECT_COUNTER_ORDER, CORRECT_IDX_TOKEN_IN_SECONDARY_INDEX, CORRECT_NON_COMPOUND_RANGE_TOMBSTONES, CORRECT_STATIC_COMPACT_IN_MC, COUNTERS, DIGEST_FOR_NULL_VALUES, DIGEST_INSENSITIVE_TO_EXPIRY, DIGEST_MULTIPARTITION_READ, EMPTY_REPLICA_PAGES, HINTED_HANDOFF_SEPARATE_CONNECTION, INDEXES, LARGE_COLLECTION_DETECTION, LARGE_PARTITIONS, LA_SSTABLE_FORMAT, LWT, MATERIALIZED_VIEWS, MC_SSTABLE_FORMAT, MD_SSTABLE_FORMAT, ME_SSTABLE_FORMAT, NONFROZEN_UDTS, PARALLELIZED_AGGREGATION, PER_TABLE_CACHING, PER_TABLE_PARTITIONERS, RANGE_SCAN_DATA_VARIANT, RANGE_TOMBSTONES, ROLES, ROW_LEVEL_REPAIR, SCHEMA_COMMITLOG, SCHEMA_TABLES_V3, SECONDARY_INDEXES_ON_STATIC_COLUMNS, SEPARATE_PAGE_SIZE_AND_SAFETY_LIMIT, STREAM_WITH_RPC_STREAM, TOMBSTONE_GC_OPTIONS, TRUNCATION_TABLE, TYPED_ERRORS_IN_READ_RPC, UDA, UDA_NATIVE_PARALLELIZED_AGGREGATION, UNBOUNDED_RANGE_TOMBSTONES, UUID_SSTABLE_IDENTIFIERS, VIEW_VIRTUAL_COLUMNS, WRITE_FAILURE_REPLY, XXHASH}, Remote common_features = {AGGREGATE_STORAGE_OPTIONS, ALTERNATOR_TTL, CDC, CDC_GENERATIONS_V2, COLLECTION_INDEXING, COMPUTED_COLUMNS, CORRECT_COUNTER_ORDER, CORRECT_IDX_TOKEN_IN_SECONDARY_INDEX, CORRECT_NON_COMPOUND_RANGE_TOMBSTONES, CORRECT_STATIC_COMPACT_IN_MC, COUNTERS, DIGEST_FOR_NULL_VALUES, DIGEST_INSENSITIVE_TO_EXPIRY, DIGEST_MULTIPARTITION_READ, EMPTY_REPLICA_PAGES, HINTED_HANDOFF_SEPARATE_CONNECTION, INDEXES, LARGE_COLLECTION_DETECTION, LARGE_PARTITIONS, LA_SSTABLE_FORMAT, LWT, MATERIALIZED_VIEWS, MC_SSTABLE_FORMAT, MD_SSTABLE_FORMAT, ME_SSTABLE_FORMAT, NONFROZEN_UDTS, PARALLELIZED_AGGREGATION, PER_TABLE_CACHING, PER_TABLE_PARTITIONERS, RANGE_SCAN_DATA_VARIANT, RANGE_TOMBSTONES, ROLES, ROW_LEVEL_REPAIR, SCHEMA_COMMITLOG, SCHEMA_TABLES_V3, SECONDARY_INDEXES_ON_STATIC_COLUMNS, SEPARATE_PAGE_SIZE_AND_SAFETY_LIMIT, STREAM_WITH_RPC_STREAM, TOMBSTONE_GC_OPTIONS, TRUNCATION_TABLE, TYPED_ERRORS_IN_READ_RPC, UDA, UDA_NATIVE_PARALLELIZED_AGGREGATION, UNBOUNDED_RANGE_TOMBSTONES, UUID_SSTABLE_IDENTIFIERS, VIEW_VIRTUAL_COLUMNS, WRITE_FAILURE_REPLY, XXHASH}
INFO  2023-07-03 07:38:26,666 [shard 0] storage_service - Host 189dd9bb-d950-446c-b527-a544925bc828/127.0.55.8 is replacing d97b63b2-72f5-4640-bba8-893420adf609/127.0.55.5
INFO  2023-07-03 07:38:26,666 [shard 0] storage_service - Replacing a node with a different IP address, my address=127.0.55.8, node being replaced=127.0.55.5
INFO  2023-07-03 07:38:26,666 [shard 0] storage_service - Save advertised features list in the 'system.local' table
INFO  2023-07-03 07:38:26,669 [shard 0] schema_tables - Schema version changed to 59adb24e-f3cd-3e02-97f0-5b395827453f
INFO  2023-07-03 07:38:26,669 [shard 0] storage_service - Starting up server gossip
INFO  2023-07-03 07:38:26,676 [shard 0] compaction - [Compact system.local 98dfab40-1974-11ee-8f2d-ec009dbc2650] Compacting [/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-nvlixxhx/test/node8/data/system/local-7ad54392bcdd35a684174e047860b377/mc-2-big-Data.db:level=0:origin=memtable,/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-nvlixxhx/test/node8/data/system/local-7ad54392bcdd35a684174e047860b377/mc-4-big-Data.db:level=0:origin=memtable]
INFO  2023-07-03 07:38:26,677 [shard 0] gossip - failure_detector_loop: Started main loop
INFO  2023-07-03 07:38:26,677 [shard 0] gossip - Waiting for 2 live nodes to show up in gossip, currently 1 present...
INFO  2023-07-03 07:38:26,686 [shard 0] compaction - [Compact system.local 98dfab40-1974-11ee-8f2d-ec009dbc2650] Compacted 2 sstables to [/jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-nvlixxhx/test/node8/data/system/local-7ad54392bcdd35a684174e047860b377/mc-6-big-Data.db:level=0]. 13kB to 7kB (~55% of original) in 6ms = 1MB/s. ~256 total partitions merged to 1.
INFO  2023-07-03 07:38:27,988 [shard 0] storage_service - Set host_id=9137e227-94f4-449e-a6c6-dceed01cff99 to be owned by node=127.0.55.4
WARN  2023-07-03 07:38:27,990 [shard 0] gossip - Fail to send EchoMessage to 127.0.55.5: seastar::rpc::closed_error (connection is closed)
WARN  2023-07-03 07:38:27,990 [shard 0] gossip - Fail to send EchoMessage to 127.0.55.6: seastar::rpc::closed_error (connection is closed)
WARN  2023-07-03 07:38:27,990 [shard 0] gossip - Fail to send EchoMessage to 127.0.55.7: seastar::rpc::closed_error (connection is closed)
INFO  2023-07-03 07:38:27,990 [shard 0] gossip - InetAddress 127.0.55.4 is now UP, status = NORMAL
INFO  2023-07-03 07:38:27,990 [shard 0] gossip - InetAddress 127.0.55.3 is now UP, status = NORMAL
INFO  2023-07-03 07:38:27,990 [shard 0] gossip - InetAddress 127.0.55.1 is now UP, status = NORMAL
INFO  2023-07-03 07:38:27,990 [shard 0] gossip - InetAddress 127.0.55.2 is now UP, status = NORMAL
INFO  2023-07-03 07:38:27,991 [shard 0] storage_service - Set host_id=c6089c7e-809f-447e-a2c8-91009de372e2 to be owned by node=127.0.55.3
INFO  2023-07-03 07:38:27,993 [shard 0] storage_service - Set host_id=66e3c3ad-d16d-469c-98e2-6c3ff5f8aaf8 to be owned by node=127.0.55.1
INFO  2023-07-03 07:38:27,994 [shard 0] gossip - Live nodes seen in gossip: {127.0.55.1, 127.0.55.2, 127.0.55.3, 127.0.55.4, 127.0.55.8}

==> [BH] it looks like the exception happened here

INFO  2023-07-03 07:38:27,995 [shard 0] init - Shutting down group 0 service
...
INFO  2023-07-03 07:38:27,997 [shard 0] storage_service - Set host_id=ec4c0157-1523-4a1d-8aeb-e56e685e507a to be owned by node=127.0.55.6

==> [BH] handle_state_normal happens in parallel to storage_service::join_token_ring

...
INFO  2023-07-03 07:38:28,026 [shard 0] gossip - InetAddress 127.0.55.6 is now DOWN, status = shutdown
...
ERROR 2023-07-03 07:38:36,614 [shard 0] init - Startup failed: std::runtime_error (Failed to parse node list: {127.0.55.6, 127.0.55.7}: invalid node=127.0.55.6: std::runtime_error (Host inet address 127.0.55.6 not found in the cluster))
```

This looks like another fallout from 50e8ec77c6332b759a1e13125771524d26b7af34.

I think that in https://github.com/scylladb/scylladb/blob/50e8ec77c6332b759a1e13125771524d26b7af34/service/storage_service.cc#L1954-L1959

`wait_for_live_nodes_to_show_up(2)` doesn't ensure that the nodes in the ignore list (that are given by ip address in this test) have their host_id set up by https://github.com/scylladb/scylladb/blob/50e8ec77c6332b759a1e13125771524d26b7af34/service/storage_service.cc#L2496-L2497

**Comment by DoronArazii** [comment]
@kbr-scylla please have a look

**Comment by bhalevy** [comment]
@DoronArazii this issue doesn't exist in 5.3 since https://github.com/scylladb/scylladb/commit/50e8ec77c6332b759a1e13125771524d26b7af34 is confined to master (5.4.dev)

**Comment by kbr-scylla** [comment]
> wait_for_live_nodes_to_show_up(2) doesn't ensure that the nodes in the ignore list (that are given by ip address in this test) have their host_id set up

How was that ensured before in this test? By ring_delay sleep?

**Comment by kbr-scylla** [comment]
We've got a bit of a chicken-and-egg problem here:
- we need to wait for NORMAL state handlers for nodes from set X
- to calculate set X, we need to calculate `ignore_nodes`
- but to calculate `ignore_nodes`, we need to wait until these nodes' NORMAL state handlers have finished (at least partially - up to the step that sets host_ids)...

Damn it, why did I think that trying to improve gossiper boot code is a good idea.

**Comment by kbr-scylla** [comment]
```cpp
    , ignore_dead_nodes_for_replace(this, "ignore_dead_nodes_for_replace", value_status::Used, "", "List dead nodes to ignore for replace operation using a comma-separated list of host IDs. E.g., scylla --ignore-dead-nodes-for-replace 8d5ed9f4-7764-4dbd-bad8-43fddce94b7c,125ed9f4-7777-1dbn-mac8-43fddce9123e")
```

it's interesting that the implementation actually supports passing IPs

**Comment by bhalevy** [comment]
> ```c++
>     , ignore_dead_nodes_for_replace(this, "ignore_dead_nodes_for_replace", value_status::Used, "", "List dead nodes to ignore for replace operation using a comma-separated list of host IDs. E.g., scylla --ignore-dead-nodes-for-replace 8d5ed9f4-7764-4dbd-bad8-43fddce94b7c,125ed9f4-7777-1dbn-mac8-43fddce9123e")
> ```
> 
> it's interesting that the implementation actually supports passing IPs

That's for backward compatibility.
We can decide to deprecate that and require passing `host_id`s.

**Comment by kbr-scylla** [comment]
However deprecation requires a deprecation period, so I still need to solve the problem.

I do have an idea, but I need to verify it.

**Comment by bhalevy** [comment]
> @bhalevy guessing this was a transient regression and doesn't need a backport

Correct.  Both patch causing the regression and fix are in 5.4 now.
